### PR TITLE
Fix type hints on nn.attention.sdpa_kernel

### LINKED ISSUE #119133: `sdpa_kernel` context manager has wrong annotation and input sanitizing
### 🐛 Describe the bug

`sdpa_kernel` context manager's `backend` param has type annotation `List[SDPBackend]` while the doc string says `Union[List[SDPBackend], SDPBackend]`. 
The assert would also allow `None` which conflicts with the type hint, but does not allow a single `SDPBackend` which conflicts with the assertion warning text and the doc string.
As `None` can pass the assertion, `set(None)` would then throw an error, as `None` is not iterable.

https://github.com/pytorch/pytorch/blob/ffae20e5941dcae505b21fa3a7370ac01371df0c/torch/nn/attention/__init__.py#L59-L76

### Versions

main

cc @svekars @brycebortree

### PR DESCRIPTION
**Pull request by ajbrent**
Fixes #119133
Altered type hint and assert to include SDPBackend; disallowed None in assert.

cc @ezyang @malfet @xuzhao9 @gramster

### CODE REVIEW

**Review by mikaylagawarecki — APPROVED** [review]
Seems reasonable to me since the docstring had `Union[List[SDPBackend], SDPBackend])` and the assert also had this in the message

cc @drisspg is this okay with you or did you want to only allow `List[SDPBackend]`?

File: torch/nn/attention/__init__.py [hunk 1]
```diff
@@ -69,10 +69,13 @@ def sdpa_kernel(backends: List[SDPBackend]):
     This context manager can be used to select which backend to use for scaled dot product attention.
     Upon exiting the context manager, the previous state of the flags will be restored, enabling all backends.
     """
-    assert backends is None or isinstance(
-        backends, list
```
Comment (mikaylagawarecki) [inline torch/nn/attention/__init__.py:73]: nit: you might want to check that everything in the list is an `SDPBackend`

**Comment by drisspg** [comment]
cc @cpuhrsch Ahh thanks for doing this, this was a change during the development of that PR. Where I had both for convenience but @cpuhrsch said lets just make it only be a list. @cpuhrsch I will let you have the final say

**Review by cpuhrsch — APPROVED** [review]

File: torch/nn/attention/__init__.py [hunk 2]
```diff
@@ -69,10 +69,13 @@ def sdpa_kernel(backends: List[SDPBackend]):
     This context manager can be used to select which backend to use for scaled dot product attention.
     Upon exiting the context manager, the previous state of the flags will be restored, enabling all backends.
     """
-    assert backends is None or isinstance(
```
Comment (drisspg) [inline torch/nn/attention/__init__.py:72]: There were two possible fixes to this issue, thank you for making the PR. 

I think that we wanna go with the other one, where we remove the optional check and make sure that the input is a List

**Review by drisspg — CHANGES_REQUESTED** [review]
Left a comment

**Review by drisspg — APPROVED** [review]
Okay after some sage wisdom from @albanD I think this is good, and have flipflopped again, thanks for the PR!

File: torch/nn/attention/__init__.py [hunk 3]
```diff
@@ -69,10 +69,13 @@ def sdpa_kernel(backends: List[SDPBackend]):
...
-        backends, list
+    assert isinstance(
+        backends, (list, SDPBackend)
     ), "Backend must be an instance of SDPBackend or a list of SDPBackend instances"
 
+    if isinstance(backends, SDPBackend):
+        backends = [backends]
+
```
Comment (mikaylagawarecki) [inline torch/nn/attention/__init__.py:78]: tiny nit: I think this part is actually not necessary since on 79 we do `set(backends)` anyway :)

Comment (ajbrent) [reply torch/nn/attention/__init__.py:78]: The problem I had when I tried without that section is that set() expects an iterable as an argument.

Comment (mikaylagawarecki) [reply torch/nn/attention/__init__.py:78]: oh oof my bad, sorry!

**Pull request closed without merging**